syntax = "proto3";

package raft;

service Raft {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc InstallSnapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);
}

message LogEntry {
    int64 term = 1;
    bytes command = 2;
}

message RequestVoteRequest {
    int64 term = 1;
    uint32 candidate_id = 2;
    int64 last_log_index = 3;
    int64 last_log_term = 4;
}

message RequestVoteResponse {
    int64 term = 1;
    bool vote_granted = 2;
}

message AppendEntriesRequest {
    int64 term = 1;
    uint32 leader_id = 2;
    int64 prev_log_index = 3;
    int64 prev_log_term = 4;
    repeated LogEntry entries = 5;
    int64 leader_commit = 6;
}

message AppendEntriesResponse {
    int64 term = 1;
    bool success = 2;
}

message InstallSnapshotRequest {
    int64 term = 1;
    uint32 leader_id = 2;
    int64 last_included_index = 3;
    int64 last_included_term = 4;
    bytes data = 5;
}

message InstallSnapshotResponse {
    int64 term = 1;
}
